### PR TITLE
search: BEAM_UOPS_MAX to prune candidates with too many uops

### PR DESCRIPTION
**Pull request by flammit**
not the highest possible performance, but adequate for most usage

TODO: will add testing to show performance before and after

### CODE REVIEW

**Comment by geohot** [comment]
Yea, we really need a test suite on this to be able to merge "tuning" PRs

**Comment by flammit** [comment]
Test commands:
```sh
# LLAMA-7B BEAM=2
time PYTHONPATH=. JIT=1 BEAM=2 CACHELEVEL=0 python3 examples/llama.py --gen 1 --prompt "Hello." --count 10 --temperature 0 --timing

# GPT2 BEAM=2
time PYTHONPATH=. JIT=1 HALF=1 BEAM=2 CACHELEVEL=0 CAST_BEFORE_VIEW=0 python3 examples/gpt2.py --count 10 --temperature 0 --timing

# CIFAR LATEBEAM=1
time PYTHONPATH=. STEPS=10 DEFAULT_FLOAT=HALF LATEBEAM=1 CACHELEVEL=0 python3 examples/hlb_cifar10.py

# CIFAR LATEBEAM=8
time PYTHONPATH=. STEPS=10 DEFAULT_FLOAT=HALF LATEBEAM=8 CACHELEVEL=0 python3 examples/hlb_cifar10.py
```

| Backend | Command | Master Total Time | Master Step Time | Branch Total Time | Branch Step Time |
| --- | --- | --- | --- | --- | --- |
| METAL | LLAMA-7B BEAM=2       | 37s | 45ms | 28s | 46ms |
| METAL | GPT2 FP16 BEAM=2      | 30s |  5ms | 16s |  5ms |
| METAL | CIFAR FP16 LATEBEAM=1 before TC fix | 5m43s | 141ms | N/A | N/A |
| METAL | CIFAR FP16 LATEBEAM=1 | 6m11s | 105-106ms | 5m06s | 114ms (!) |
| METAL | CIFAR FP16 LATEBEAM=1 BEAM_MIN_PROGRESS=2 | N/A | N/A | 5m03s | 111-112ms (!) |
| METAL | CIFAR FP16 LATEBEAM=1 BEAM_UPCAST_MAX=256 | N/A | N/A | 6m12s | 100-102ms |
| METAL | CIFAR FP16 LATEBEAM=8 | 54m | 97ms | 26m | 110ms (!) |
| METAL | CIFAR FP16 LATEBEAM=8 BEAM_UPCAST_MAX=256 | N/A | N/A | 33m | 98ms |
| HSA   | LLAMA-7B BEAM=2       | 66s | 22ms | 52s | 22ms |
| HSA   | GPT2 FP16 BEAM=2      | 67s | 3.7ms | 51s | 3.7ms |
| HSA   | CIFAR FP16 LATEBEAM=1 | 20m | 35-40ms | 6m40s | 36-40ms |
| HSA   | CIFAR FP16 LATEBEAM=1 BEAM_MIN_PROGRESS=2 | N/A | N/A | 6m53s | 39-41ms |
| HSA   | CIFAR FP16 LATEBEAM=1 BEAM_UPCAST_MAX=256 | N/A | N/A | 8m01s | 39-41ms |
| HSA   | CIFAR FP16 LATEBEAM=8 | 43m | 33ms | 18m | 30ms |
| HSA   | CIFAR FP16 LATEBEAM=8 BEAM_UPCAST_MAX=256 | N/A | N/A | 24m | 37ms |
| CUDA  | LLAMA-7B BEAM=2       | 70-80s | 19ms | 72s | 19ms |
| CUDA  | GPT2 FP16 BEAM=2      | 70s | 1.7ms | 58s | 1.7ms |
| CUDA  | CIFAR FP16 LATEBEAM=1 | 6m | 36-38ms | 6m | 33-37ms |
| CUDA  | CIFAR FP16 LATEBEAM=1 BEAM_MIN_PROGRESS=2 | N/A | N/A | 6m25s | 33-34ms |
| CUDA  | CIFAR FP16 LATEBEAM=1 BEAM_UPCAST_MAX=256 | N/A | N/A | 6m00s | 30-32ms |
| CUDA  | CIFAR FP16 LATEBEAM=8 | 33m | 29ms | 33m | 30ms |
| CUDA  | CIFAR FP16 LATEBEAM=8 BEAM_UPCAST_MAX=256 | N/A | N/A | OOM (not tiny only 32gbs) | OOM |

**Comment by flammit** [comment]
Added one more attempt at CIFAR: `CUDA=1 LATEBEAM=8 BEAM_UPCAST_MAX=256 BEAM_MAX_TASKS_PER_CHILD=1` but it didn't finish after more than 17hrs.

**Comment by chenyuxyz** [comment]
maybe a version of `BEAM_MIN_PROGRESS` that relies on relative time can mitigate the slowdown issue.

**Comment by flammit** [comment]
My mistake - it wasn't running CUDA=1... my drivers didn't load after reboot after reinstalling and rerunning, it finished in 43 minutes.

**Comment by chenyuxyz** [comment]
can we get a version with `BEAM_MAX_TASKS_PER_CHILD` change and uops MAX only first? i think these 2 are the least controversial

**Comment by flammit** [comment]
Done!  also added sanity line to .gitignore as discussed

File: tinygrad/features/search.py [hunk 1]
```diff
@@ -45,9 +45,10 @@ def _time_program(variables:List[Variable], outcount:int, rdev:Compiled, lib:byt
     if early_stop is not None and early_stop < tms[-1]: break
   return tms
 
-def _compile_linearizer(compiler:Compiler, lin:Linearizer, name:Optional[str]=None) -> Tuple[bytes, Optional[List[int]], Optional[List[int]],
-                                                                                             List[Variable], int, float, int]:
+def _compile_linearizer(compiler:Compiler, lin:Linearizer, name:Optional[str]=None, enforce_max:bool=False) \
+  -> Tuple[bytes, Optional[List[int]], Optional[List[int]], List[Variable], int, float, int]:
   lin.linearize()
+  if (uops_max:=getenv("BEAM_UOPS_MAX", 3000)) > 0 and len(lin.uops.uops) >= uops_max and enforce_max: raise Exception("too many uops")
```
Comment (chenyuxyz) [inline tinygrad/features/search.py:51]: pick a more specific error than plain `Exception`

also put the predicate first, i would write something like
`if enforce_max and len(lin.uops.uops) >= getenv("BEAM_UOPS_MAX", 3000) > 0: raise ...`

**Review by chenyuxyz — COMMENTED** [review]
how did you pick 3000 as the default threshold, are we hitting it currently for the usual examples?

**Comment by chenyuxyz** [comment]
also the benchmark beam runs took 50%-100% longer

**Comment by flammit** [comment]
> how did you pick 3000 as the default threshold, are we hitting it currently for the usual examples?

it was just a heuristic based on looking at `BEAM_LOG=1` outputs as a proxy for 1-ish second compile times (here's one from a tiny red): 

Screenshot 2024-04-15 at 1 33 38 PM

File: tinygrad/features/search.py [hunk 2]
```diff
@@ -112,7 +113,7 @@ def beam_search(lin:Linearizer, rawbufs:List[Buffer], amt:int, allow_test_size=T
 
   default_parallel, min_progress_micros = 1 if lin.opts.device in {"CUDA", "HSA", "KFD"} else 0, getenv("BEAM_MIN_PROGRESS",0.01)
   if beam_pool is None and getenv("PARALLEL", default_parallel):
-    beam_pool = multiprocessing.get_context("spawn").Pool(multiprocessing.cpu_count(), _init_worker, (), getenv("BEAM_MAX_TASKS_PER_CHILD", 16))
```
Comment (chenyuxyz) [inline tinygrad/features/search.py:115]: 16 has been working fine. my bad for going back to this but can we change this in a separate pr?

File: tinygrad/features/search.py (hunk 1 above)
Comment (chenyuxyz) [reply tinygrad/features/search.py:51]: and this

**Comment by chenyuxyz** [comment]
will measure resnet compile time again after this change

**Pull request merged**